### PR TITLE
Make clients closeable

### PR DESCRIPTION
**Pull request by AsafMah**
#### Pull Request Description

Make clients (both clients and ingest clients) closeable, so users could free their resources when done using them.
---

#### Future Release Comment
_[Add description of your change, to include in the next release]_
_[Delete any or all irrelevant sections, e.g. if your change does not warrant a release comment at all]_

**Breaking Changes:**
- None

**Features:**
- Make clients (both clients and ingest clients) closeable, so users could free their resources when done using them.

**Fixes:**
- None

### CODE REVIEW

File: kusto/ingest/managed.go [hunk 1]
```diff
@@ -146,3 +146,15 @@ func (m *Managed) newProp() properties.All {
...
 }
+
+func (m *Managed) Close() error {
+	var err error
+	err = m.queued.Close()
+	err2 := m.streaming.Close()
+	if err == nil {
+		err = err2
+	} else {
+		err = fmt.Errorf("combined error: %v %v", err, err2)
```
Comment (yogilad) [inline kusto/ingest/managed.go:157]: Same (quotations, seperator)

File: kusto/kusto.go [hunk 2]
```diff
@@ -431,3 +434,19 @@ func (*Client) contextSetup(ctx context.Context, mgmtCall bool) (context.Context
...
+	var err error
+	if c.conn != nil {
+		err = c.conn.Close()
+	}
+	if c.ingestConn != nil {
+		err2 := c.ingestConn.Close()
+		if err == nil {
+			err = err2
+		} else {
+			err = fmt.Errorf("combined error: %v %v", err, err2)
```
Comment (yogilad) [inline kusto/kusto.go:448]: Same

Comment (yogilad) [reply kusto/kusto.go:448]: Also, note if you nest these you'll have multiple "combined error:" prefixes

File: kusto/ingest/internal/conn/conn.go [hunk 3]
```diff
@@ -179,3 +187,15 @@ func copyHeaders(header http.Header) http.Header {
 	}
 	return headers
 }
+
+func (c *Conn) Close() error {
+	for {
```
Comment (yogilad) [inline kusto/ingest/internal/conn/conn.go:192]: You do not need the for loop

Comment (AsafMah) [reply kusto/ingest/internal/conn/conn.go]: Fixed

File: kusto/kusto.go (hunk 2 above)
Comment (AsafMah) [reply kusto/kusto.go]: Fixed

File: kusto/ingest/managed.go (hunk 1 above)
Comment (AsafMah) [reply kusto/ingest/managed.go]: Fixed

File: kusto/ingest/ingest.go [hunk 4]
```diff
@@ -240,3 +241,18 @@ func (i *Ingestion) newProp() properties.All {
...
+func (i *Ingestion) Close() error {
+	i.mgr.Close()
+	var err error
+	err = i.fs.Close()
+	if i.streamConn != nil {
+		err2 := i.streamConn.Close()
+		if err == nil {
+			err = err2
+		} else {
+			err = fmt.Errorf("combined error: %v %v", err, err2)
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/ingest/ingest.go]: Fixed

**Review by alonadam — APPROVED** [review]

**Review by yogilad — APPROVED** [review]

**Pull request merged**